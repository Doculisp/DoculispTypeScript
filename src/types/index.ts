// Export all types for external consumption
export * from './types.ast';
export * from './types.astDoculisp';
export * from './types.astProject';
export * from './types.containers';
export * from './types.controller';
export * from './types.document';
export * from './types.docWriter';
export * from './types.fileHandler';
export * from './types.filePath';
export * from './types.general';
export * from './types.includeBuilder';
export * from './types.internal';
export * from './types.sringBuilder';
export * from './types.stringWriter';
export * from './types.structure';
export * from './types.textHelpers';
export * from './types.tokens';
export * from './types.trimArray';
export * from './types.variableTable';
export * from './types.version';