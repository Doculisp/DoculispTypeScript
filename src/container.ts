import { IContainer, IDependencyContainer, IDependencyManager, IDictionary, IRegisterable, ITestableContainer, Valid } from "./types.containers";
import { globSync } from 'glob';

function findModules(container: IContainer) {
    let moduleNames = (
        globSync('./dist/**/*.js')
        .map(p => {
            return p.replace('dist', '.').replace('\\src', '').replace('.js', '').replaceAll('\\', '/');
        })
    );

    moduleNames.forEach(n => {
        import(n).then(m => {
            const keys = Object.keys(m);
            keys.
                filter(key => {
                    let value = m[key];
                    return !!value.name && !!value.builder;
                }).
                forEach(key => container.register(m[key]));
        });
    });
}

class Container implements ITestableContainer {
    _registry: IDictionary<IRegisterable> = {};
    _replacements: IDictionary<IRegisterable> = {};
    _cache: IDictionary<Valid<any>> = {};
    _repCache: IDictionary<Valid<any>> = {};
    _testable = false;

    constructor();
    constructor(registry: IDictionary<IRegisterable>);
    constructor(registry?: IDictionary<IRegisterable>) {
        this._testable = !!registry;
        if(!!registry) {
            const keys = Object.keys(registry);
            keys.forEach(key => this._registry[key] = registry[key]);
        }
        else {
            findModules(this);
        }
    }

    _hasKey(registry: IDictionary<IRegisterable>, key: string): boolean {
        return !!registry[key];
    }

    _register(registry: IDictionary<IRegisterable>, registerable: IRegisterable): ITestableContainer {
        let key = registerable.name;
        if (this._hasKey(registry, key)) {
            throw new Error(`Module named "${key}" already registered.`);
        }
        registry[key] = registerable;
        return this
    }

    register(registerable: IRegisterable): ITestableContainer {
        if (!registerable.name) {
            throw new Error('Must have a valid name to be registered.');
        }
        this._register(this._registry, registerable);
        return this;
    }
    
    registerValue(value: any, name?: string | undefined): ITestableContainer {
        if (!value.name && !name) {
            throw new Error('Most provide a name as a property or a parameter when registering a value as a module.');
        }

        const registerable: IRegisterable = {
            builder: () => { return value; },
            name: name ?? value.name,
            singleton: true
        };

        return this.register(registerable);
    }
    
    registerBuilder(fn: (...args: any[]) => any, dependencies: string[], name?: string | undefined, singleton?: boolean | undefined): ITestableContainer {
        if(!fn.name && !name) {
            throw new Error('Must provide a name on the function or as a parameter to register a builder.');
        }

        const registerable: IRegisterable = {
            builder: fn,
            name: name ?? fn.name,
            dependencies: dependencies ?? [],
            singleton: !!singleton,
        }
        
        this.register(registerable);
        return this;
    }

    _build<T>(moduleName: string, modules: string[]): Valid<T> {
        let moduleBuilder = this._replacements[moduleName];
        let replaced = true;
        if (!moduleBuilder) {
            moduleBuilder = this._registry[moduleName];
            replaced = false;
        }

        if (!moduleBuilder) {
            let t: any = undefined;
            try {
                t = require(moduleName);
            }
            catch {
            }

            if (!t) {
                throw new Error(`No module called "${moduleName}" registered`);
            }
            this.registerBuilder(() => require(moduleName), [], moduleName);
            moduleBuilder = this._registry[moduleName] as IRegisterable;
        }

        if(this._repCache[moduleName]) {
            return this._repCache[moduleName];
        }
        
        if(!replaced && this._cache[moduleName]) {
            return this._cache[moduleName];
        }

        let recursive = 
            moduleBuilder.dependencies?.filter(name => modules.includes(name)) ?? [];

        modules[modules.length] = moduleName;

        if (0 < recursive.length) {
            let moduleNames = modules.map(name => `"${name}"`).join(' => ');
            let recursiveModules = recursive.map(name => `"${name}"`).join(', ');
            throw new Error(`Circular dependencies between (${moduleNames} => [${recursiveModules}])`);
        }

        let dependencies: Valid<any>[] = 
            moduleBuilder.dependencies?.map((name) => this._build(name, modules)) ?? [];

        let result: Valid<T> = moduleBuilder.builder.apply(moduleBuilder.builder, dependencies);

        if(moduleBuilder.singleton) {
            if(replaced) {
                this._repCache[moduleName] = result;
            }
            else {
                this._cache[moduleName] = result;
            }
        }

        return result;
    }
    
    build(moduleName: string): Valid<any> {
        return this._build<any>(moduleName, []);
    }
    
    buildAs<T>(moduleName: string): Valid<T> {
        return this._build<T>(moduleName, []);
    }
    
    buildTestable(): ITestableContainer {
        return new Container(this._registry);
    }
    
    supportsReplace(): boolean {
        return this._testable;
    }

    _replace(registerable: IRegisterable, isPackage?: boolean) : ITestableContainer {
        let key = registerable.name;
        if(!isPackage && !this._hasKey(this._registry, key)) {
            throw new Error(`Cannot replace module "${key}" as it has not been registered.`);
        }
        if(!isPackage && this._hasKey(this._replacements, key)) {
            throw new Error (`Cannot replace module "${key}" as it has not been registered.`)
        }

        return this._register(this._replacements, registerable);
    }
    
    replace(registerable: IRegisterable): ITestableContainer {
        return this._replace(registerable);
    }
    
    replaceBuilder(fn: (...args: any[]) => any, dependencies: string[], name?: string | undefined, singleton?: boolean | undefined): ITestableContainer {
        if (!fn.name && !name) {
            throw new Error('Must provide a name either on the passed method or parameter to replace builder.');
        }

        let registerable: IRegisterable = {
            builder: fn,
            dependencies: dependencies ?? [],
            name: name ?? fn.name,
            singleton: !!singleton,
        };
        
        return this._replace(registerable);// 
    }
    
    replaceValue(value: any, name?: string | undefined): ITestableContainer {
        let key = name ?? value.name;
        if (!key) {
            throw new Error('Cannot replace value unless it has a name property or the name is passed as a parameter.');
        }

        return this.replaceBuilder(() => { return value; }, [], key, true);
    }
    
    replacePackageBuilder(fn: (...args: any[]) => any, name?: string | undefined, singleton?: boolean | undefined): ITestableContainer {
        if(!fn.name && !name) {
            throw new Error('Must provide a name either on the passed method or parameter to replace package builder.');
        }

        let registerable: IRegisterable = {
            builder: fn,
            name: name ?? fn.name,
            singleton: !!singleton,
        };
        
        return this._replace(registerable, true);
    }
    
    replacePackageValue(value: any, name?: string | undefined): ITestableContainer {
        if(!value.name && !name) {
            throw new Error('Cannot replace package with value unless it has a name property or the name is passed as a parameter.');
        }

        return this.replacePackageBuilder(() => value, name ?? value.name);
    }
    
    restore(moduleName: string): ITestableContainer {
        delete this._replacements[moduleName];
        delete this._repCache[moduleName];
        return this;
    }
    
    restoreAll(): ITestableContainer {
        const keys = Object.keys(this._replacements);
        let that = this;
        keys.forEach(key => that.restore(key));
        return this;
    }
    
    getModuleList(): string[] {
        const keys = Object.keys(this._registry);
        return Object.assign([], keys);
    }
}

const core = new Container();
const registry = core as IDependencyContainer;
const manager = core as IDependencyManager;
const container = core as IContainer;

export {
    registry,
    manager,
    container,
};