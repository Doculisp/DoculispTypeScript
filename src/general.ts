import { IRegisterable } from "./types.containers";
import { IFail, ILocation, ISuccess, IUtil } from "./types.general";

class Location implements ILocation {
    private readonly _line: number;
    private readonly _char: number;
    
    constructor(line: number, char: number) {
        this._line = line;
        this._char = char;
    }

    public get line(): number {
        return this._line;
    }

    public get char(): number {
        return this._char;
    }

    toString(): string {
        return `{ line: ${this._line}, char: ${this._char} }`;
    }

    asJson() {
        return { line: this._line, char: this._char };
    }
}

function buildGeneral(): IUtil {
    function ok<T>(successfulValue: T) : ISuccess<T> {
        return {
            value: successfulValue,
            success: true,
        };
    };

    function fail(message: string, documentPath: string) : IFail {
        return {
            message,
            documentPath,
            success: false,
        };
    };

    function location(line: number, char: number): ILocation {
        return new Location(line, char);
    }

    return {
        ok,
        fail,
        location,
    };
}

const astParser: IRegisterable = {
    builder: () => buildGeneral(),
    name: 'util',
    singleton: true,
    dependencies: []
};

export {
    astParser,
};