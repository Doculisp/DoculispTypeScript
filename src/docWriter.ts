import { AstBulletStyle, AstPart, IAst, IHeader, ILoad, ISectionWriter, ITableOfContents, ITitle, IWrite } from "./types.ast";
import { IRegisterable } from "./types.containers";
import { IDocWriter } from "./types.docWriter";
import { IUtil, Result } from "./types.general";

function buildWriter(util: IUtil): IDocWriter {
    function getLabel(style: AstBulletStyle, load: ILoad): string {
        switch (style) {
            case 'bulleted':
            case 'numbered':
            case 'unlabeled':
                return '';
            default:
                return load.sectionLabel;
        }
    }

    function getNumberedBullet(): (label: string, text: string) => string {
        let cnt = 1;
        function bullet(label: string, text: string): string {
            const lbl = 0 < label.length ? `${label}: ` : '';
            const result = `${cnt}. ${lbl}${text}`;
            cnt++;
            return result;
        }

        return bullet;
    }

    function getBullet(): (label: string, text: string) => string {
        function bullet(label: string, text: string): string {
            const lbl = 0 < label.length ? `${label}: ` : '';
            return `* ${lbl}${text}`;
        }

        return bullet;
    }

    function getNoBullet(): (label: string, text: string) => string {
        function bullet(label: string, text: string): string {
            const lbl = 0 < label.length ? `${label}: ` : '';
            return `${lbl}${text}`;
        }
    }

    function getBulletBuilder(style: AstBulletStyle): (label: string, text: string) => string {
        switch (style) {
            case 'bulleted':
            case 'bulleted-labeled':
                return getBullet();
            case 'numbered':
            case 'numbered-labeled':
                return getNumberedBullet();
            default:
                return getNoBullet();
        }
    }

    function compileToc(astToc: ITableOfContents, loads: ILoad): string {
        if(astToc.bulletStyle === 'no-table') {
            return '';
        }
    }

    function compileContentLocation(loads: ILoad[]): string {
        const parts: string[] = [];

        loads.forEach(load => {
            if(load.document) {
                const section = compileSection(load.document);
                parts.push(section);
            }
        });

        return `${parts.join('\n')}`;
    }

    function compileHeader(astHeader: IHeader): string {
        const padding = '#'.repeat(astHeader.depthCount);
        return `${padding} ${astHeader.text} ${padding}`;
    }

    function compileTitle(astTitle: ITitle): string {
        return astTitle.label;
    }

    function compileWrite(astWrite: IWrite): string {
        return astWrite.value;
    }

    function compilePart(astPart: AstPart) : string {
        astPart.
    }

    function compileSection(astSection: ISectionWriter): string {
        const parts: string[] = [];

        
    }

    function compile(astResult: Result<IAst>): Result<string> {
        if(!astResult.success) {
            return astResult;
        }

        const ast = astResult.value;

        if(ast.section.type === 'ast-empty') {
            return util.ok('');
        }

        throw new Error("Not yet Implemented");
    }

    function write(path: string, astResult: Result<IAst>): Result<string> {
        throw new Error('Not yet Implemented');
    }

    return {
        compile,
        write,
    };
}

const documentWriter: IRegisterable = {
    builder: (util: IUtil) => buildWriter(util),
    name: 'docWriter',
    singleton: true,
    dependencies: ['util', 'fileHandler']
};

export {
    documentWriter,
};