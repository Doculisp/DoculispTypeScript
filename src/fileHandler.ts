import { IRegisterable } from "./types.containers";
import { IFileHandler } from "./types.fileHandler";

import { IUtil, Result } from "./types.general";

function buildLoader(util: IUtil, fs: any): IFileHandler {
    function load(path: string): Result<string> {
        try {
            const value: string = fs.readFileSync(path, {encoding: 'utf8'});
            return util.ok(value);
        } catch (error) {
            return util.fail(`${error}`, path);
        }
    }

    return {
        load
    };
}

const loader: IRegisterable = {
    builder: (util: IUtil, fs: any) => buildLoader(util, fs),
    name: 'fileHandler',
    dependencies: ['util', 'fs'],
    singleton: true
};

export {
    loader,
};