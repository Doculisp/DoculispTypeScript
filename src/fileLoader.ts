import { IRegisterable } from "./types.containers";
import { IFileLoader } from "./types.fileLoader";

import { IUtil, Result } from "./types.general";

function buildLoader(util: IUtil, fs: any): IFileLoader {
    function load(path: string): Result<string> {
        try {
            const value: string = fs.readFileSync(path, {encoding: 'utf8'});
            return util.ok(value);
        } catch (error) {
            return util.fail(`${error}`, path);
        }
    }

    return {
        load
    };
}

const loader: IRegisterable = {
    builder: (util: IUtil, fs: any) => buildLoader(util, fs),
    name: 'fileLoader',
    dependencies: ['util', 'fs'],
    singleton: true
};

export {
    loader,
};