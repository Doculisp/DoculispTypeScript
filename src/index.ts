#! /usr/bin/env node
import { container } from './container';
import { Command } from 'commander';
import { Result } from './types/types.general';
import { IController } from './types/types.controller';
import { IVersion } from './types/types.version';
import figlet from 'figlet';

const program = new Command();
const controller = container.buildAs<IController>('controller');
const versionGetter = container.buildAs<IVersion>('version');

const versionMaybe = versionGetter.getVersion();


let version = '?.?.?'

if(!versionMaybe.success) {
    console.log('doculisp cannot determine its version')
} else {
    version = versionMaybe.value;
}

const helptext = '\n\n' + figlet.textSync('doculisp', {
    font: 'Cybermedium'
}) + `\n                     Version: ${version}`;

program.addHelpText('beforeAll', helptext);
program.addHelpText('afterAll', '\n\n');

program
    .name('doculisp')
    .description('A compiler for markdown')
    .version(version)
    .requiredOption('-s, --source <source_path>', 'the source file to compile')
    .requiredOption('-d, --output <output_path>', 'the output document path for the compiled markdown')
    .option('-t, --test', 'runs the compiler without generating the output file.')
    .parse(process.argv);

function reportResult(result: Result<string | false>) {
    if (result.success) {
        const message = !!result.value ? `Ok! ${result.value} successfully written` : 'Ok!';
        console.log(message);
    }
    else {
        console.log(JSON.stringify(result, null, 4));
    }
}

const options = program.opts();

const sourcePath: string = options['source'];

if (options['test']) {
    const result = controller.test(sourcePath);
    reportResult(result);
}
else {
    const outputPath: string = options['output'];
    const result = controller.compile(sourcePath, outputPath);
    reportResult(result);
}
