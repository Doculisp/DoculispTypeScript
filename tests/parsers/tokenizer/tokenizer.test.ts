import { container } from "../../../src/container";
import { configure } from "approvals/lib/config";
import { getVerifier } from "../../tools";
import { Options } from "approvals/lib/Core/Options";
import { ITestableContainer } from "../../../src/types.containers";
import { TokenFunction } from '../../../src/types.tokens';
import { IFail, ILocation, ISuccess, IUtil, Result } from "../../../src/types.general";
import { DocumentMap } from "../../../src/types.document";

describe('tokenizer', () => {
    let environment: ITestableContainer = undefined as any;
    let tokenizer: TokenFunction = undefined as any;
    let verifyAsJson: (data: any, options?: Options) => void = undefined as any;
    let ok: (successfulValue: any) => ISuccess<any> = undefined as any;
    let fail: (message: string, documentPath: string) => IFail = undefined as any;
    let util: IUtil = undefined as any;

    beforeAll(() => {
        verifyAsJson = getVerifier(configure);
    });

    beforeEach(() => {
        environment = container.buildTestable();
        tokenizer = environment.buildAs<TokenFunction>('tokenizer');
        util = environment.buildAs<IUtil>('util');
        ok = util.ok;
        fail = util.fail;
    });

    test('should fail if document parsing failed', () => {
        const parseResult = fail('This document did not parse', 'X:/non-exist.dlisp') as Result<DocumentMap>;

        const result = tokenizer(parseResult);

        verifyAsJson(result);
    });

    test('should return empty if given an empty parse result', () => {
        const parseResult: Result<DocumentMap> = ok({
            projectLocation: { documentPath: 'c:/empty/readme.md', documentDepth: 0, documentIndex: 0 },
            parts: [],
        });

        const result = tokenizer(parseResult);

        verifyAsJson(result);
    });

    test('should tokenize text as text', () => {
        const parseResult: Result<DocumentMap> = ok({
            projectLocation: { documentPath: 'D:/comments/simple.md', documentDepth: 0, documentIndex: 0 },
            parts: [
                {
                    type: 'text',
                    text: 'hello my text',
                    location: util.location('D:/comments/simple.md', 0, 0, 5, 23),
                }
            ],
        });

        const result = tokenizer(parseResult);

        verifyAsJson(result);
    });

    describe('handling Doculisp', () => {
        test('should tokenize an empty comment', () => {
            let parseResult: Result<DocumentMap> = ok({
                projectLocation: { documentPath: 'D:/comments/simple.md', documentDepth: 0, documentIndex: 0 },
                parts: [
                    {
                        type: 'lisp',
                        text: '(*)',
                        location: util.location('D:/comments/simple.md', 0, 0, 2, 1),
                    },
                ],
            });
            
            let result = tokenizer(parseResult);

            verifyAsJson(result);
        });
        
        test('should tokenize an single atom', () => {
            const start: ILocation = util.location('D:/comments/simple.md', 0, 0, 4, 2);
            let parseResult: Result<DocumentMap> = ok({
                projectLocation: { documentPath: 'D:/comments/simple.md', documentDepth: 0, documentIndex: 0 },
                parts: [
                    {
                        type: 'lisp',
                        text: '(atom)',
                        location: start,
                    },
                ],
            });
            
            const result = tokenizer(parseResult);

            verifyAsJson(result);
        });
        
        test('should tokenize an single atom with space after atom', () => {
            const start: ILocation = util.location('D:/comments/simple.md', 0, 0, 4, 2);
            let parseResult: Result<DocumentMap> = ok({
                projectLocation: { documentPath: 'D:/comments/simple.md', documentDepth: 0, documentIndex: 0 },
                parts: [
                    {
                        type: 'lisp',
                        text: '(atom )',
                        location: start,
                    },
                ],
            });
            
            const result = tokenizer(parseResult);

            verifyAsJson(result);
        });
        
        test('should tokenize an single atom with new line after atom', () => {
            const start: ILocation = util.location('D:/comments/simple.md', 0, 0, 4, 2);
            let parseResult: Result<DocumentMap> = ok({
                projectLocation: { documentPath: 'D:/comments/simple.md', documentDepth: 0, documentIndex: 0 },
                parts: [
                    {
                        type: 'lisp',
                        text: '(atom\r\n)',
                        location: start,
                    },
                ],
            });
            
            const result = tokenizer(parseResult);

            verifyAsJson(result);
        });
        
        test('should tokenize an single atom containing only numbers', () => {
            const start: ILocation = util.location('D:/comments/simple.md', 0, 0, 4, 2 );
            let parseResult: Result<DocumentMap> = ok({
                projectLocation: { documentPath: 'D:/comments/simple.md', documentDepth: 0, documentIndex: 0 },
                parts: [
                    {
                        type: 'lisp',
                        text: '(123987)',
                        location: start,
                    },
                ],
            });
            
            const result = tokenizer(parseResult);

            verifyAsJson(result);
        });
        
        test('should tokenize an single atom with hyphen and underscore', () => {
            const start: ILocation = util.location('D:/comments/simple.md', 0, 0, 4, 2);
            let parseResult: Result<DocumentMap> = ok({
                projectLocation: { documentPath: 'D:/comments/simple.md', documentDepth: 0, documentIndex: 0 },
                parts: [
                    {
                        type: 'lisp',
                        text: '(atom-start_end)',
                        location: start,
                    },
                ],
            });
            
            const result = tokenizer(parseResult);

            verifyAsJson(result);
        });

        test('should tokenize a single atom with a single word parameter', () => {
            const start: ILocation = util.location('Z:/parameter.md', 0, 0, 1, 13);

            let parseResult: Result<DocumentMap> = ok({
                projectLocation: { documentPath: 'Z:/parameter.md', documentDepth: 0, documentIndex: 0 },
                parts: [
                    {
                        type: 'lisp',
                        text: '(the thing)',
                        location: start,

                    }
                ],
            }); 

            const result = tokenizer(parseResult);

            verifyAsJson(result);
        });

        test('should tokenize a single atom with a multi word parameter', () => {
            const start: ILocation = util.location('Z:/parameter.md', 0, 0, 1, 13);

            let parseResult: Result<DocumentMap> = ok({
                projectLocation: { documentPath: 'Z:/parameter.md', documentDepth: 0, documentIndex: 0 },
                parts: [
                    {
                        type: 'lisp',
                        text: '(title the thing from beyond\n\tthe swamp)',
                        location: start,

                    }
                ],
            }); 

            const result = tokenizer(parseResult);

            verifyAsJson(result);
        });

        test('should handle nested lisp', () => {
            const start: ILocation = util.location('A:/main.md', 0, 0, 2, 1);

            let parseResult: Result<DocumentMap> = ok({
                projectLocation: { documentPath: 'A:/main.md', documentDepth: 0, documentIndex: 0 },
                parts: [
                    {
                        type: 'lisp',
                        location: start,
                        text: `(section-meta
        (external
            (Section ./structure.md)
        )
    )`,
                    }
                ]
            });

            const result = tokenizer(parseResult);

            verifyAsJson(result);
        });

        test('should handle comment with nested lisp', () => {
            const start: ILocation = util.location('A:/main.md', 0, 0, 2, 1);

            let parseResult: Result<DocumentMap> = ok({
                projectLocation: { documentPath: 'A:/main.md', documentDepth: 0, documentIndex: 0 },
                parts: [
                    {
                        type: 'lisp',
                        location: start,
                        text: `(section-meta
        (*external
            (Section ./structure.md)
            (*Section ./comments.md)
            (Section ./toc.md)
        )
    )`,
                    }
                ]
            });

            const result = tokenizer(parseResult);

            verifyAsJson(result);
        });
    });
});