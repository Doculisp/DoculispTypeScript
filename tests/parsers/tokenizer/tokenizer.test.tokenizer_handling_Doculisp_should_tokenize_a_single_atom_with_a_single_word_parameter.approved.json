{
  "success": true,
  "value": {
    "projectLocation": {
      "documentDepth": 0,
      "documentIndex": 0,
      "documentPath": "Z:/parameter.md"
    },
    "tokens": [
      {
        "location": {
          "char": 13,
          "documentDepth": 0,
          "documentIndex": 0,
          "documentPath": "Z:/parameter.md",
          "line": 1
        },
        "type": "token - open parenthesis"
      },
      {
        "location": {
          "char": 14,
          "documentDepth": 0,
          "documentIndex": 0,
          "documentPath": "Z:/parameter.md",
          "line": 1
        },
        "text": "the",
        "type": "token - atom"
      },
      {
        "location": {
          "char": 18,
          "documentDepth": 0,
          "documentIndex": 0,
          "documentPath": "Z:/parameter.md",
          "line": 1
        },
        "text": "thing",
        "type": "token - parameter"
      },
      {
        "location": {
          "char": 23,
          "documentDepth": 0,
          "documentIndex": 0,
          "documentPath": "Z:/parameter.md",
          "line": 1
        },
        "type": "token - close parenthesis"
      }
    ]
  }
}
